### PR TITLE
Support for np.ptp

### PR DESCRIPTION
**Pull request by rjenc29**
Initial commit for CI.

### CODE REVIEW

**Comment by rjenc29** [comment]
I think the CI failure is unrelated (a timeout).  

I've not attempted to handle complex dtype, although numpy does. 

The implementation looks a bit duplicitous, but my initial attempts to refactor either (1) led to accidental type casting int -> float and / or (2) degraded performance, so I have stuck with plan A. 

Anyway, if you have time to review and let me know feedback, that would be great.

**Comment by rjenc29** [comment]
Added support for complex dtype in the last commit - won't make any more changes pending review feedback. ta!

File: numba/tests/test_array_reductions.py [hunk 1]
```diff
@@ -665,6 +668,93 @@ def test_nancumsum_basic(self):
...
+
+        def a_variations():
+            yield np.arange(10)
+            yield np.array([-1.1, np.nan, 2.2])
+            yield np.array([-np.inf, 5])
+            yield (4, 2, 5)
+            yield (1,)
+            yield np.full(5, 5)
+            yield [2.2, -2.3, 0.1]
+            a = np.linspace(-10, 10, 16).reshape(4, 2, 2)
```
Comment (stuartarchibald) [inline numba/tests/test_array_reductions.py]: Perhaps also reverse `[::-1]`, and `np.random.shuffle` this to get some randomized values for input?

File: numba/tests/test_array_reductions.py [hunk 2]
```diff
@@ -665,6 +668,93 @@ def test_nancumsum_basic(self):
...
+        def make_array(real_nan=False, imag_nan=False):
+            real = np.linspace(-4, 4, 25)
+            if real_nan:
+                real[4:9] = np.nan
+            imag = np.linspace(-5, 5, 25)
+            if imag_nan:
+                imag[7:12] = np.nan
+            return (real + 1j * imag).reshape(5, 5)
+
+        comp = make_array()
```
Comment (stuartarchibald) [inline numba/tests/test_array_reductions.py]: Use a loop and `itertools.product([True, False], repeat=2)` ?

File: numba/tests/test_array_reductions.py [hunk 3]
```diff
@@ -665,6 +668,93 @@ def test_nancumsum_basic(self):
...
+        cfunc = jit(nopython=True)(pyfunc)
+
+        # Exceptions leak references
+        self.disable_leak_check()
+
+        with self.assertTypingError() as e:
+            cfunc(np.array((True, True, False)))
+
+        msg = "Boolean dtype is unsupported (as per NumPy)"
+        assert msg in str(e.exception)
```
Comment (stuartarchibald) [inline numba/tests/test_array_reductions.py]: use `self.assertIn()`?  (and below)

File: numba/tests/test_array_reductions.py [hunk 4]
```diff
@@ -665,6 +668,93 @@ def test_nancumsum_basic(self):
...
+
+        real = np.ones(8)
+        imag = np.arange(-4, 4)
+        comp = real + 1j * imag
+        check(comp)
+        comp = real - 1j * imag
+        check(comp)
+
+        comp = np.full((4, 4), fill_value=(1 - 1j))
+        check(comp)
```
Comment (stuartarchibald) [inline numba/tests/test_array_reductions.py]: Is there any testing of scalar complex values?

**Comment by stuartarchibald** [comment]
RE: https://github.com/numba/numba/pull/3572#issuecomment-446105533

> The implementation looks a bit duplicitous, but my initial attempts to refactor either (1) led to accidental type casting int -> float and / or (2) degraded performance, so I have stuck with plan A.

Does this https://github.com/numba/numba/commit/b5d5ba836f844c661f594a1ca53ba59586ee33f0 help?! Not sure how it performs in relation to the current implementation. If it's useful and performs well you are more than welcome to use it (or parts of) as you see fit!

**Comment by rjenc29** [comment]
Thanks for the feedback - all very sensible - I will action the points raised as soon as poss.
Have a great xmas!

File: numba/tests/test_array_reductions.py (hunk 1 above)
Comment (rjenc29) [reply numba/tests/test_array_reductions.py]: Done

File: numba/tests/test_array_reductions.py (hunk 3 above)
Comment (rjenc29) [reply numba/tests/test_array_reductions.py]: Done

File: numba/tests/test_array_reductions.py (hunk 2 above)
Comment (rjenc29) [reply numba/tests/test_array_reductions.py]: Done

File: numba/tests/test_array_reductions.py (hunk 4 above)
Comment (rjenc29) [reply numba/tests/test_array_reductions.py]: I'm not sure what you mean.

**Comment by rjenc29** [comment]
I will have a proper look at b5d5ba8 over the festive period and run some benchmarks, but it looks neat.

**Comment by rjenc29** [comment]
b5d5ba8 looks good to me - I made a couple of small tweaks and added a couple of additional test cases in the last commit - let me know what you think.  ta!

File: numba/tests/test_array_reductions.py (hunk 4 above)
Comment (stuartarchibald) [reply numba/tests/test_array_reductions.py]: I think I've just spotted some I missed above, was looking for e.g. `x + y * j` as input.

**Comment by stuartarchibald** [comment]
@rjenc29 thanks for the fixes, they seem good. Thanks also taking a look at the de-duplication code, did the performance change much/at all?

@seibert please could you take a look at this, now contains a chunk of code I wrote!

**Comment by rjenc29** [comment]
The performance seemed about the same (or marginally better), for the collection of benchmarks I used - certainly the difference was small enough to accurately measure :)

**Comment by seibert** [comment]
Should we raise a ValueError like NumPy does when a zero-length array is passed in?

**Comment by rjenc29** [comment]
In theory, the implementation should hit a code branch which does indeed do that, per this bit of `test_ptp_exceptions`:

```python
with self.assertRaises(ValueError) as e:
    cfunc(np.array([]))
 
msg = "zero-size array reduction not possible"
self.assertIn(msg, str(e.exception))
```

**Comment by seibert** [comment]
ah, missed that.  Thanks!  This looks ready to merge.

**Comment by stuartarchibald** [comment]
Thanks for the additional review @seibert.

**Pull request merged**